# Data Cleaning - FIFA 21 players dataset

![FIFA 21](https://www.fifa-infinity.com/wp-content/uploads/2020/03/fifa-21-covid.jpg)

---

## Import Libraries & Load Data

In [1]:
# Import libraries for data manipulation
import numpy as np
import pandas as pd

# for getting month name
import calendar as cal

In [2]:
# load data
df = pd.read_csv("fifa21 raw data v2.csv")

C:\Users\User\AppData\Local\Temp\ipykernel_12268\2042770981.py:2: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("fifa21 raw data v2.csv")


In [3]:
# setting maximum number of columns to display to 100
pd.set_option("display.max_columns",100)

In [4]:
# viewing the first 5 rows of the dataset
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioni

In [5]:
# viewing the last 5 rows of the dataset
df.tail()

ID          Name        LongName  \
18974  247223        Xia Ao          Ao Xia   
18975  258760      B. Hough       Ben Hough   
18976  252757   R. McKinley  Ronan McKinley   
18977  243790  Wang Zhen'ao    Zhen'ao Wang   
18978  252520     Zhou Xiao       Xiao Zhou   

                                               photoUrl  \
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                      Club     Contract Positions Height  \
18974    47   55        \n\n\n\nWuhan Zall  2018 ~ 2022        CB  178cm   
18975    47   67   \n\n\n\nOldham Athletic  2020 ~ 2021        CM  175cm   
18976    47   65        \n\n\n\nDerry City  2019 ~ 2020        CM  179cm   
18977    47   57  \n\n\n\nDalian YiFang FC  2020 ~ 2022        RW  175cm   
18978    47   57  \n\n\n\nDalian YiFang FC  2019 ~ 2023    CB, LB  188cm   

      Weight Preferred Foot  BOV Best Position        Joined Loan Date End  \
18974   66kg          Right   49            CB  Jul 13, 2018           NaN   
18975   65kg          Right   51           CAM   Aug 1, 2020           NaN   
18976   74kg          Right   49           CAM   Mar 8, 2019           NaN   
18977   69kg          Right   48            ST  Sep 22, 2020           NaN   
18978   75kg           Left   50            LB  Jul 29, 2019           NaN   

       Value  Wage Release Clause  Attacking  Crossing  Finishing  \
18974  €100K   €1K           €70K        145        23         26   
18975  €130K  €500          €165K        211        38         42   
18976  €120K  €500          €131K        200        30         34   
18977  €100K   €2K           €88K        215        45         52   
18978  €100K   €1K           €79K        163        40         18   

       Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
18974                43             26       27    142         27     23   
18975                40             56       35    219         46     40   
18976                43             54       39    207         43     39   
18977                34             42       42    194         51     35   
18978                40             45       20    171         40     33   

       FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
18974           21            29            42       294            68   
18975           35            50            48       305            63   
18976           31            47            47       290            59   
18977           31            31            46       254            62   
18978           20            40            38       279            60   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
18974            60       69         46       51    221          36       57   
18975            64       61         51       66    226          48       58   
18976            66       51         47       67    242          45       52   
18977            55       50         33       54    235          56       45   
18978            63       57         47       52    231          37       66   

       Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
18974       54        50          24        192          48             50   
18975       43        47      

---

## Data Exploration

In [6]:
df.shape

(18979, 77)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

In [8]:
# checking for duplicates
df.duplicated().sum()

0

In [9]:
# checking for missing values
df.isnull().sum()

ID              0
Name            0
LongName        0
photoUrl        0
playerUrl       0
             ... 
PAS             0
DRI             0
DEF             0
PHY             0
Hits         2595
Length: 77, dtype: int64

In [10]:
#statistical summary
df.describe().T

count           mean           std    min       25%  \
ID                18979.0  226403.384794  27141.054157   41.0  210135.0   
Age               18979.0      25.194109      4.710520   16.0      21.0   
↓OVA              18979.0      65.718636      6.968999   47.0      61.0   
POT               18979.0      71.136414      6.114635   47.0      67.0   
BOV               18979.0      66.751726      6.747193   48.0      62.0   
Attacking         18979.0     248.938142     74.299428   42.0     222.0   
Crossing          18979.0      49.688392     18.131153    6.0      38.0   
Finishing         18979.0      45.842405     19.567081    3.0      30.0   
Heading Accuracy  18979.0      51.942726     17.294409    5.0      44.0   
Short Passing     18979.0      58.768112     14.519106    7.0      54.0   
Volleys           18979.0      42.696507     17.646937    3.0      30.0   
Skill             18979.0     256.479214     78.650601   40.0     222.0   
Dribbling         18979.0      55.587491     18.761314    5.0      49.0   
Curve             18979.0      47.246957     18.207790    4.0      35.0   
FK Accuracy       18979.0      42.391011     17.227947    5.0      31.0   
Long Passing      18979.0      52.725381     15.178151    5.0      43.0   
Ball Control      18979.0      58.528373     16.565892    5.0      54.0   
Movement          18979.0     317.718689     55.879046  122.0     289.0   
Acceleration      18979.0      64.357553     14.889981   13.0      57.0   
Sprint Speed      18979.0      64.412930     14.638743   12.0      57.0   
Agility           18979.0      63.366721     14.596277   14.0      55.0   
Reactions         18979.0      61.655619      9.072114   24.0      56.0   
Balance           18979.0      63.925865     14.062285   12.0      56.0   
Power             18979.0     296.605195     50.732390  122.0     264.0   
Shot Power        18979.0      57.801676     13.308747   18.0      48.0   
Jumping           18979.0      64.577375     11.844870   15.0      58.0   
Stamina           18979.0      62.653986     15.804223   12.0      55.0   
Strength          18979.0      64.786922     12.488672   16.0      57.0   
Long Shots        18979.0      46.785236     19.300534    4.0      32.0   
Mentality         18979.0     254.199747     64.595613   50.0     227.5   
Aggression        18979.0      55.591285     17.140414    9.0      44.0   
Interceptions     18979.0      46.394489     20.698078    3.0      25.0   
Positioning       18979.0      50.298962     19.428701    2.0      40.0   
Vision            18979.0      53.857105     13.708574    9.0      45.0   
Penalties         18979.0      48.057906     15.655999    6.0      39.0   
Composure         18979.0      58.073660     12.079110   12.0      50.0   
Defending         18979.0     139.909268     61.219472   20.0      83.0   
Marking           18979.0      46.559197     20.139324    3.0      29.0   
Standing Tackle   18979.0      47.705622     21.367690    5.0      27.0   
Sliding Tackle    18979.0      45.644449     20.922087    4.0      25.0   
Goalkeeping       18979.0      81.436693     84.665579   10.0      48.0   
GK Diving         18979.0      16.416302     17.554194    2.0       8.0   
GK Handling       18979.0      16.207124     16.816305    2.0       8.0   
GK Kicking        18979.0      16.076453     16.491103    2.0       8.0   
GK Positioning    18979.0      16.217187     17.002239    2.0       8.0   
GK Reflexes       18979.0      16.519627     17.854079    2.0       8.0   
Total Stats       18979.0    1595.286949    269.874789  747.0    1452.0   
Base Stats        18979.0     355.702197     40.761117  232.0     327.0   
PAC               18979.0      67.453975     10.677859   25.0      61.0   
SHO               18979.0      53.457031     13.827425   16.0      44.0   
PAS               18979.0      57.681016     10.081857   25.0      51.0   
DRI               18979.0      62.875020      9.927415   25.0      57.0   
DEF               18979.0      49.866221     

**Observations:**
- There are no duplicates.
- `Hits` has 2595 missing values.
- `Loan Date End` has 17966 missing values. 
- Some columns have the wrong datatypes.
- Some values have random characters in them.

We will deal with the above issues in the next section.

---

## Data Cleaning

### Transforming `Height` and `Weight` 

In [11]:
# checking datatypes
print("Data type of Height: ", df['Height'].dtype)
print("Data type of Weight: ", df['Weight'].dtype)

Data type of Height:  object
Data type of Weight:  object


Both are of the data type 'object'. We need to convert it to numeric data. 

**`Height`**

In [12]:
# checking unique values of 'Height'
df["Height"].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

The values have 'cm' in them and some values are in inches.

Let us first remove 'cm' from the values.

In [13]:
# removing 'cm'
df['Height'] = df['Height'].str.replace('cm', '')


df["Height"].unique()

array(['170', '187', '188', '181', '175', '184', '191', '178', '193',
       '185', '199', '173', '168', '176', '177', '183', '180', '189',
       '179', '195', '172', '182', '186', '192', '165', '194', '167',
       '196', '163', '190', '174', '169', '171', '197', '200', '166',
       '6\'2"', '164', '198', '6\'3"', '6\'5"', '5\'11"', '6\'4"',
       '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"', '5\'4"',
       '201', '158', '162', '161', '160', '203', '157', '156', '202',
       '159', '206', '155'], dtype=object)

Now let us convert feet and inches into cm.

We will use the below conversion:

1 ft = 30.48cm

1 in = 2.54cm

In [14]:
# function for converting feet and inches to cm
def feet_inches_to_cm(height):
    if '\'' in height:
        feet, inches = height.split('\'')
        return round(int(feet) * 30.48 + int(inches.strip('"')) * 2.54)
    else:
        return int(height)

df['Height'] = df['Height'].apply(feet_inches_to_cm)

In [15]:
df['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 201,
       158, 162, 161, 160, 203, 157, 156, 202, 159, 206, 155], dtype=int64)

**`Weight`**

In [16]:
# checking unique values of 'Weight'
df.Weight.unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

Some values are in pounds and some in kg. Let us make it all consistent and convert to kg.

First, let us remove 'kg' from the values

In [17]:
# removing 'kg'
df['Weight'] = df['Weight'].str.replace('kg', '')

df["Weight"].unique()

array(['72', '83', '87', '70', '68', '80', '71', '91', '73', '85', '92',
       '69', '84', '96', '81', '82', '75', '86', '89', '74', '76', '64',
       '78', '90', '66', '60', '94', '79', '67', '65', '59', '61', '93',
       '88', '97', '77', '62', '63', '95', '100', '58', '183lbs',
       '179lbs', '172lbs', '196lbs', '176lbs', '185lbs', '170lbs',
       '203lbs', '168lbs', '161lbs', '146lbs', '130lbs', '190lbs',
       '174lbs', '148lbs', '165lbs', '159lbs', '192lbs', '181lbs',
       '139lbs', '154lbs', '157lbs', '163lbs', '98', '103', '99', '102',
       '56', '101', '57', '55', '104', '107', '110', '53', '50', '54',
       '52'], dtype=object)

Now let us convert pound to kg.

We will use the below conversion:

1 lbs = 0.45359237 kg

In [18]:
# converting pounds to kg
def pound_to_kg(weight):
    if 'lbs' in weight:
        weight = weight.strip('lbs') # remove 'lbs' from the values
        return round(int(weight) * 0.45359237)
    else:
        return int(weight)

df['Weight'] = df['Weight'].apply(pound_to_kg)

In [19]:
df['Weight'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52], dtype=int64)

Let us check the datatypes again.

In [20]:
print("Data type of Height: ", df['Height'].dtype)
print("Data type of Weight: ", df['Weight'].dtype)

Data type of Height:  int64
Data type of Weight:  int64


All the values are consistent and in the same units. 

Finally, let us change the column names for more clarity.

In [21]:
# renaming the columns
df.rename(columns={'Height': 'Height(cm)', 'Weight':'Weight(kg)'}, inplace=True)

### Transforming `Joined` 

In [22]:
# viewing the values in 'Joined'
df['Joined']


0         Jul 1, 2004
1        Jul 10, 2018
2        Jul 16, 2014
3        Aug 30, 2015
4         Aug 3, 2017
             ...     
18974    Jul 13, 2018
18975     Aug 1, 2020
18976     Mar 8, 2019
18977    Sep 22, 2020
18978    Jul 29, 2019
Name: Joined, Length: 18979, dtype: object

In [23]:
# checking datatype
df['Joined'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 18979 entries, 0 to 18978
Series name: Joined
Non-Null Count  Dtype 
--------------  ----- 
18979 non-null  object
dtypes: object(1)
memory usage: 148.4+ KB


Let us split this into Day, Month, and Year columns.

Firts, lets change the datatype to datetime.

In [24]:
df['Joined'] = pd.to_datetime(df['Joined'])

In [25]:
df['Joined']

0       2004-07-01
1       2018-07-10
2       2014-07-16
3       2015-08-30
4       2017-08-03
           ...    
18974   2018-07-13
18975   2020-08-01
18976   2019-03-08
18977   2020-09-22
18978   2019-07-29
Name: Joined, Length: 18979, dtype: datetime64[ns]

Now, lets split the dates into **day**, **month**, **year**.

In [26]:
# creating the new columns
df['Join Day'] = df['Joined'].dt.day
df['Join Month'] = df['Joined'].dt.month
df['Join Year'] = df['Joined'].dt.year

In [27]:
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions  Height(cm)  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF         170   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW         187   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK         188   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM         181   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM         175   

   Weight(kg) Preferred Foot  BOV Best Position     Joined Loan Date End  \
0          72           Left   93            RW 2004-07-01           NaN   
1          83          Right   92            ST 2018-07-10           NaN   
2          87          Right   91            GK 2014-07-16           NaN   
3          70          Right   91           CAM 2015-08-30           NaN   
4          68          Right   91            LW 2017-08-03           NaN   

     Value   Wage Release Clause  Attacking  Crossing  Finishing  \
0  €103.5M  €560K        €138.4M        429        85         95   
1     €63M  €220K         €75.9M        437        84         95   
2    €120M  €125K        €159.4M         95        13         11   
3    €129M  €370K          €161M        407        94         82   
4    €132M  €270K        €166.5M        408        85         87   

   Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
0                70             91       88    470         96     93   
1                90             82       86    414         88     81   
2                15             43       13    109         12     13   
3                55             94       82    441         88     85   
4                62             87       87    448         95     88   

   FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0           94            91            96       451            91   
1           76            77            92       431            87   
2           14            40            30       307            43   
3           83            93            92       398            77   
4           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353   

The `Join Month` column represents numeric values of each month. Lets change this to month names.

In [28]:
# function to get month name
def get_month_name(month_number):
    return cal.month_name[month_number]

# applying the function
df['Join Month'] = df['Join Month'].apply(get_month_name)

In [29]:
df['Join Month'].unique()

array(['July', 'August', 'January', 'October', 'September', 'May', 'June',
       'February', 'April', 'December', 'March', 'November'], dtype=object)

In [30]:
# checking datatypes
print('Day: ', df['Join Day'].dtype)
print('Month: ', df['Join Month'].dtype)
print('Year: ', df['Join Year'].dtype)

Day:  int64
Month:  object
Year:  int64


### Transforming `Value`, `Wage`, `Release Clause` 

In [31]:
# checking datatypes of each column
print('Value: ', df['Value'].dtype)
print('Wage: ', df['Wage'].dtype)
print('Release Clause: ', df['Release Clause'].dtype)

Value:  object
Wage:  object
Release Clause:  object


These columns should be of numeric datatypes as it represents market value, weekly wage, and release clause of each player.

To do this, we will first need to remove the `€` symbol from the values.

In [32]:
# removing '€' symbol
df['Value'] = df['Value'].str.replace('€','')
df['Wage'] = df['Wage'].str.replace('€','')
df['Release Clause'] = df['Release Clause'].str.replace('€','')

Now we can change the columns to numeric datatype.

In [33]:
# function for converting string to number
def str_to_num(value):
    if 'K' in value:
        return round(float(value.strip('K')) * 1000)
    elif 'M' in value:
        return round(float(value.strip('M')) * 1000000)
    else:
        return int(value)

df['Value'] = df['Value'].apply(str_to_num)
df['Wage'] = df['Wage'].apply(str_to_num)
df['Release Clause'] = df['Release Clause'].apply(str_to_num)


In [34]:
# checking datatypes of each column
print('Value: ', df['Value'].dtype)
print('Wage: ', df['Wage'].dtype)
print('Release Clause: ', df['Release Clause'].dtype)

Value:  int64
Wage:  int64
Release Clause:  int64


### Transforming `Club` 

In [35]:
df.Club.unique()

array(['\n\n\n\nFC Barcelona', '\n\n\n\nJuventus',
       '\n\n\n\nAtlético Madrid', '\n\n\n\nManchester City',
       '\n\n\n\nParis Saint-Germain', '\n\n\n\nFC Bayern München',
       '\n\n\n\nLiverpool', '\n\n\n\nReal Madrid', '\n\n\n\nChelsea',
       '\n\n\n\nTottenham Hotspur', '\n\n\n\nInter', '\n\n\n\nNapoli',
       '\n\n\n\nBorussia Dortmund', '\n\n\n\nManchester United',
       '\n\n\n\nArsenal', '\n\n\n\nLazio', '\n\n\n\nLeicester City',
       '\n\n\n\nBorussia Mönchengladbach', '\n\n\n\nReal Sociedad',
       '\n\n\n\nAtalanta', '\n\n\n\nOlympique Lyonnais', '\n\n\n\nMilan',
       '\n\n\n\nVillarreal CF', '\n\n\n\nRB Leipzig', '\n\n\n\nCagliari',
       '\n\n\n\nAjax', '\n\n\n\nSL Benfica', '\n\n\n\nAS Monaco',
       '\n\n\n\nWolverhampton Wanderers', '\n\n\n\nEverton',
       '\n\n\n\nFiorentina', '\n\n\n\nFC Porto', '\n\n\n\nRC Celta',
       '\n\n\n\nTorino', '\n\n\n\nSevilla FC', '\n\n\n\nGrêmio',
       '\n\n\n\nReal Betis', '\n\n\n\nRoma', '\n\n\n\nNewcastle Unite

The `Club` column has newline characters in its values. So let us remove those.

In [36]:
# removing newline characters
df['Club'] = df['Club'].str.replace('\n', '')

### Creating new column `'Agreement'`

Let us now create a new column `'Agreement'` based on the column 'Contract'. 'Contract' contains different types of values representing paid agreements, loan agreements, and free agent statuses for different players. Let us group these into **'Contract'**, **'Loan'**, and **'Free'**.

In [37]:
# function for the condition
def contract_type(contract):
    if '~' in contract:
        type = 'Contract'
    elif 'Loan' in contract:
        type = 'Loan'
    else:
        type = 'Free'
    return type
        
df['Agreement'] = df['Contract'].apply(contract_type)

In [38]:
df.Agreement.unique()

array(['Contract', 'Loan', 'Free'], dtype=object)

### Transforming `W/F`, `SM`, `IR`

**W/F ("Weak Foot")**: It refers to the rating or assessment of a player's weaker foot, indicating their proficiency in using their non-dominant foot for activities such as passing, shooting, and dribbling. The rating is often represented as a number between 1 and 5, with 5 being the highest proficiency.

**SM ("Skill Moves")**: "Skill Moves" refer to the range of advanced techniques and tricks that a player can perform in the game to outmaneuver opponents. The "Skill Moves" rating usually ranges from 1 to 5 or 1 to 5 stars, indicating the complexity and effectiveness of the player's skill moves repertoire.

**IR (International Reputation)**: "International Reputation" refers to the perceived prestige and recognition of a player on the international stage. It typically reflects the player's reputation based on their performance in international competitions and their standing among football fans worldwide. The "International Reputation" rating ranges from 1 to 5 or 1 to 5 stars, with higher values indicating greater recognition and prestige.

`W/F`, `SM`, and `IR` ranges between 1 - 5. These values also have a "★" character in it. Let us remove this character from the columns. 

Let us also change the column names for improved clarity and readability.

In [39]:
# removing '★'
df['W/F'] = df['W/F'].str.replace('★', '')
df['SM'] = df['SM'].str.replace('★', '')
df['IR'] = df['IR'].str.replace('★', '')

In [40]:
# removing white spaces
df['W/F'] = df['W/F'].str.strip()
df['SM'] = df['SM'].str.strip()
df['IR'] = df['IR'].str.strip()

In [41]:
# changing datatype to numeric
df['W/F'] = pd.to_numeric(df['W/F'])
df['SM'] = pd.to_numeric(df['SM'])
df['IR'] = pd.to_numeric(df['IR'])

In [42]:
# changing column names
df.rename(columns={'W/F': 'Weak Foot', 'SM': 'Skill Moves', 'IR': 'Intl Rep'}, inplace=True)

In [43]:
# checking values & datatypes
print('Weak Foot')
print(df['Weak Foot'].unique())
print(df['Weak Foot'].dtype)

print('----------------------')

print('Skill Moves')
print(df['Skill Moves'].unique())
print(df['Skill Moves'].dtype)

print('----------------------')

print('Intl Rep')
print(df['Intl Rep'].unique())
print(df['Intl Rep'].dtype)

print('----------------------')

Weak Foot
[4 3 5 2 1]
int64
----------------------
Skill Moves
[4 5 1 2 3]
int64
----------------------
Intl Rep
[5 3 4 2 1]
int64
----------------------


### Transforming `Hits`

**Hits** may refer to successful passes or tackles made by players during a match. In this case, “hits” might be used to indicate successful actions.

In [44]:
df.Hits.unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

In [45]:
# checking missing values
df.Hits.isnull().sum()

2595

The `Hits` column is currently not numeric. It contains 2595 missing values. Some of the values contain suffix 'K' in them which needs to be changed.

In [46]:
# function to convert value to number
def to_num(value):
    if pd.isnull(value): # check if value is NaN and return 0 if it is
        return 0
    elif isinstance(value, float):  # Check if value is already a float
        return int(value)  # Convert float to int
    elif 'K' in value:
        return round(float(value.strip('K')) * 1000) # removing 'K' suffix and multiplying the value with 1000
    else:
        return int(value)

# applying the function
df['Hits'] = df['Hits'].apply(to_num)

In [47]:
df['Hits'].unique()

array([ 771,  562,  150,  207,  595,  248,  246,  120, 1600,  130,  321,
        189,  175,   96,  118,  216,  212,  154,  205,  202,  339,  408,
        103,  332,   86,  173,  161,  396, 1100,  433,  242,  206,  177,
       1500,  198,  459,  117,  119,  209,   84,  187,  165,  203,   65,
        336,  126,  313,  124,  145,  538,  182,  101,   45,  377,   99,
        194,  403,  414,  593,  374,  245, 3200,  266,  299,  309,  215,
        265,  211,  112,  337,   70,  159,  688,  116,   63,  144,  123,
         71,  224,  113,  168,   61,   89,  137,  278,   75,  148,  176,
        197,  264,  214,  247,  402,  440, 1700, 2300,  171,  320,  657,
         87,  259,  200,  255,  253,  196,   60,   97,   85,  169,  256,
        132,  239,  166,  121,  109,   32,   46,  122,   48,  527,  199,
        282,   51, 1900,  642,  155,  323,  288,  497,  509,   79,   49,
        270,  511,   80,  128,  115,  156,  204,  143,  140,  152,  220,
        134,  225,   94,   74,  135,  142,   50,   

### Transforming `Loan Date End`

The `Loan Date End` column represents the end date of a player's loan. It has some missing values. This relates to players who have not taken a loan. We will replace the missing values with **"No Loan"**.

In [48]:
df['Loan Date End'].fillna("No Loan", inplace = True)

# checking for missing values
df['Loan Date End'].isnull().sum()

0

   ### Removing `photoUrl` and `playerUrl` columns

We will be dropping these columns as they do not add any value to the analysis.

In [49]:
df.drop(columns = ['photoUrl', 'playerUrl'], inplace = True)

df.head()

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   ↓OVA  POT                 Club     Contract   Positions  Height(cm)  \
0    93   93         FC Barcelona  2004 ~ 2021  RW, ST, CF         170   
1    92   92             Juventus  2018 ~ 2022      ST, LW         187   
2    91   93      Atlético Madrid  2014 ~ 2023          GK         188   
3    91   91      Manchester City  2015 ~ 2023     CAM, CM         181   
4    91   91  Paris Saint-Germain  2017 ~ 2022     LW, CAM         175   

   Weight(kg) Preferred Foot  BOV Best Position     Joined Loan Date End  \
0          72           Left   93            RW 2004-07-01       No Loan   
1          83          Right   92            ST 2018-07-10       No Loan   
2          87          Right   91            GK 2014-07-16       No Loan   
3          70          Right   91           CAM 2015-08-30       No Loan   
4          68          Right   91            LW 2017-08-03       No Loan   

       Value    Wage  Release Clause  Attacking  Crossing  Finishing  \
0  103500000  560000       138400000        429        85         95   
1   63000000  220000        75900000        437        84         95   
2  120000000  125000       159400000         95        13         11   
3  129000000  370000       161000000        407        94         82   
4  132000000  270000       166500000        408        85         87   

   Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
0                70             91       88    470         96     93   
1                90             82       86    414         88     81   
2                15             43       13    109         12     13   
3                55             94       82    441         88     85   
4                62             87       87    448         95     88   

   FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0           94            91            96       451            91   
1           76            77            92       431            87   
2           14            40            30       307            43   
3           83            93            92       398            77   
4           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353          63             29   
2       41        78          12        140          34             19   
3       89        74          91        408          76             66   
4       81        50          84        356          51             36   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0           93      95         75         96         91       32   
1           95      82         84         95         84       28   
2           11      65         11         68         57       27   
3           88      94         84         91        186       68   
4           87      90         92         93         94       35   

   Standing Tackle  Sliding Tack

### Changing column names

In [52]:
df.rename(columns = 
          {'LongName': 'Full Name', 
          '↓OVA': 'Overall Rating', # overall rating of the player's skills and abilities
          'POT': 'Potential Rating', #  potential rating of the player's skills and abilities
          'FK Accuracy': 'Free Kick Accuracy',
           'A/W': 'Attacking Work Rate',
           'D/W': 'Defensive Work Rate',
           'PAS': 'Passing', # player’s “Passing” attribute
           'PHY': 'Physical' # player’s “Physical” attribute
          }, 
          inplace = True)

df.head()

ID               Name                     Full Name Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   Overall Rating  Potential Rating                 Club     Contract  \
0              93                93         FC Barcelona  2004 ~ 2021   
1              92                92             Juventus  2018 ~ 2022   
2              91                93      Atlético Madrid  2014 ~ 2023   
3              91                91      Manchester City  2015 ~ 2023   
4              91                91  Paris Saint-Germain  2017 ~ 2022   

    Positions  Height(cm)  Weight(kg) Preferred Foot  BOV Best Position  \
0  RW, ST, CF         170          72           Left   93            RW   
1      ST, LW         187          83          Right   92            ST   
2          GK         188          87          Right   91            GK   
3     CAM, CM         181          70          Right   91           CAM   
4     LW, CAM         175          68          Right   91            LW   

      Joined Loan Date End      Value    Wage  Release Clause  Attacking  \
0 2004-07-01       No Loan  103500000  560000       138400000        429   
1 2018-07-10       No Loan   63000000  220000        75900000        437   
2 2014-07-16       No Loan  120000000  125000       159400000         95   
3 2015-08-30       No Loan  129000000  370000       161000000        407   
4 2017-08-03       No Loan  132000000  270000       166500000        408   

   Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
0        85         95                70             91       88    470   
1        84         95                90             82       86    414   
2        13         11                15             43       13    109   
3        94         82                55             94       82    441   
4        85         87                62             87       87    448   

   Dribbling  Curve  Free Kick Accuracy  Long Passing  Ball Control  Movement  \
0         96     93                  94            91            96       451   
1         88     81                  76            77            92       431   
2         12     13                  14            40            30       307   
3         88     85                  83            93            92       398   
4         95     88                  89            81            95       453   

   Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  \
0            91            80       91         94       95    389          86   
1            87            91       87         95       71    444          94   
2            43            60       67         88       49    268          59   
3            77            76       78         91       76    408          91   
4            94            89       96         91       83    357          80   

   Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  \
0       68       72        69          94        347          44   
1       95       84        78          93        353          63   
2       78       41        78          12        140          34   
3       63       89        74          91        408          76   
4       62       81        50          84        356          51   

   Interceptions  Positioning  Vision  Penalties  Composure  Defending  \
0             40           93      95         75         96         91   
1             29           95      82         84         95         84   
2             19           11      65         11         68         57   
3             66           88      94         84       

---

## Exporting the cleaned data

In [53]:
#creating an excel file of the updated dataset

df.to_excel('FIFA_21_Cleaned.xlsx', sheet_name ='Cleaned Data')

---